atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultKeyValueSnapshotReader.java

```diff
@@ -0,0 +1,365 @@
+
+                // This means that this transaction started too long ago. When we do garbage collection,
+                // we clean up old values, and this transaction started at a timestamp before the garbage collection.
+                readSentinelHandler.handleReadSentinel(key);
+            } else {
+                long theirCommitTimestamp =
+                        commitTimestamps.getIfAbsent(value.getTimestamp(), TransactionConstants.FAILED_COMMIT_TS);
+                if (theirCommitTimestamp == TransactionConstants.FAILED_COMMIT_TS) {
+                    keysToReload.put(key, value.getTimestamp());
+
+                    // This is from a failed transaction so we can roll it back and then reload it.
+                    keysToDelete.put(key, value.getTimestamp());
+                    //                    getCounter(AtlasDbMetricNames.CellFilterMetrics.INVALID_COMMIT_TS, tableRef)
+                    //                            .inc();
+                } else if (theirCommitTimestamp > startTimestampSupplier.getAsLong()) {
+                    // The value's commit timestamp is after our start timestamp.
+                    // This means the value is from a transaction which committed
+                    // after our transaction began. We need to try reading at an
+                    // earlier timestamp.
+                    keysToReload.put(key, value.getTimestamp());
+                    //
+                    // getCounter(AtlasDbMetricNames.CellFilterMetrics.COMMIT_TS_GREATER_THAN_TRANSACTION_TS, tableRef)
+                    //                            .inc();
+                } else {
+                    // The value has a commit timestamp less than our start timestamp, and is visible and valid.
+                    if (value.getContents().length != 0) {
+                        resultsCollector.add(Maps.immutableEntry(key, transformer.apply(value)));
+                        keysAddedBuilder.add(key);
+                    }
+                }
+            }
+        }
+        Set<Cell> keysAddedToResults = keysAddedBuilder.build();
+
+        if (!keysToDelete.isEmpty()) {
+            // if we can't roll back the failed transactions, we should just try again
+            if (!rollbackFailedTransactions(tableRef, keysToDelete, commitTimestamps, transactionService)) {
+                return Futures.immediateFuture(getRemainingResults(rawResults, keysAddedToResults));
+            }
+        }
+
+        if (!keysToReload.isEmpty()) {
+            // TODO (jkong): Do we still need to do this given we validate the output? Check with @lmeireles
+            return Futures.transform(
+                    cellGetter.get(tableRef, keysToReload),
+                    nextRawResults -> {
+                        boolean allPossibleCellsReadAndPresent = nextRawResults.size() == keysToReload.size();
+                        preCommitConditionValidator.validatePreCommitRequirementsOnReadIfNecessary(
```

Comment:
I presume this needs to happen here? Is there a way we could hoist it up to SnapshotTransaction? Or that would potentially abort transactions a bit later than we'd like?

Reply:
Could e.g. SnapshotReader return something from the final batch to indicate whether this should happen?

Reply:
It probably can actually be hoisted, yeah (see the TODO comment a few lines up). This is basically the end of the method, and it's ok for this to have some strange abnormalities as long as by the time we get back to the user everything's good.

Reply:
... Actually, in the interim I'd rather not change semantics here as there could be unintended consequences. Streamlining the checks could be done as something later post discussions with @LucasIME and possibly other stakeholders as well, but I'd rather isolate this change as a separate thing.